Add servo calibration page (#31)
NUbots/NUbook: src/book/03-guides/03-hardware/02-servo-calibration.mdx

2. If there are multiple servos discovered, the new servo should be the option with ID = 1 and Baud Rate = 57600 bps.

3. Update the following parameters (don't forget to press save after each update):

   - ID: Varies depending on where the servo will be placed. See below for guidance.

     | Servo Position       | ID  |
     | -------------------- | --- |
     | Right Shoulder Pitch | 1   |
     | Left Shoulder Pitch  | 2   |
     | Right Shoulder Roll  | 3   |
     | Left Shoulder Roll   | 4   |
     | Right Elbow          | 5   |
     | Left Elbow           | 6   |
     | Right Hip Yaw        | 7   |
     | Left Hip Yaw         | 8   |
     | Right Hip Roll       | 9   |
     | Left Hip Roll        | 10  |
     | Right Hip Pitch      | 11  |
     | Left Hip Pitch       | 12  |
     | Right Knee           | 13  |
     | Left Knee            | 14  |
     | Right Ankle Pitch    | 15  |
     | Left Ankle Pitch     | 16  |
     | Right Ankle Roll     | 17  |
     | Left Ankle Roll      | 18  |
     | Head Yaw             | 19  |
     | Head Pitch           | 20  |

>      Note that roll is about the x-axis, pitch is about the y-axis, and yaw is about the z-axis.

Reviewer: Adding in a reference that is not accurate to the current coordinates, but will become accurate once Tom finishes updating the coordinates conventions for the servos

   - Baud Rate (Bus): 1M bps = 2M/(1 + 1)

     ![Screenshot of Dynamixel Wizard Baud Rate selection with 1M bps selected](./images/servo_calibration/update_parameters.png)

## Testing

To test the servo, perform the following instructions:

1. Set the Goal Position in Dynamixel Wizard to 0, +90 and -90 and observe the position of the arrow on the horn. It should match the picture.

> 2. Set the Goal Position in Dynamixel Wizard to some position either side of +180 to test if the servo is in joint mode. If it is it should take the long route around to the second position. If it does not, it is not in joint mode. _instructions on how to fix_

Reviewer 2: I don't understand `_instructions on how to fix_` ? is that meant to be a  link? A header for the next section perhaps?

Reviewer 3: Probably a "insert instructions here" placeholder.

4. Disconnect the servo from the USB attachment for a few seconds and then reconnect. If the software does not detect the disconnection, retry. If it fails, redo the calibration from step one. The software will then install the firmware to the servo.

5. Rotate the servo so that the dot is in the 6-o'clock position (as seen in the instructions). Use the tool pictured to move the horn and use the marks on the tool to line up the dot correctly.

   ![Photo of the servo rotating tool](./images/servo_calibration/servo_tool.jpg)

6. Continue to rotate to each of the 3-o'clock, 12-o'clock and 9-o'clock positions when told by the software.

## Parameters

To update the servo parameters, perform the following instructions:

1. Select "**Scan**" and wait for the attached servo to be discovered.

   Optional: brand new servos should be preset to an ID of 1. Lower the ID range in "**Options > Scan**" to speed up discovery.
   ![Screenshot of Dynamixel Wizard performing a scan with no servos discovered](./images/servo_calibration/scan.png)
   ![Screenshot of Dynamixel Wizard performing a scan with 3 servos discovered](./images/servo_calibration/discover.png)

> 2. If there are multiple servos discovered, the new servo should be the option with ID = 1 and Baud Rate = 57600 bps.

Reviewer 4: Might be worth mentioning this in the optional bit above to help people speed up the search and narrow down potential results

     <details>
        <summary>Detailed diagram of NUgus servo axes</summary>

     ![Diagram of NUgus robot with roll, pitch and yaw axes defined for each servo.](./images/servo_calibration/nugus_servo_axis.svg)
     </details>

   - Baud Rate (Bus): 1M bps = 2M/(1 + 1)

     ![Screenshot of Dynamixel Wizard Baud Rate selection with 1M bps selected](./images/servo_calibration/update_parameters.png)

## Testing

To test the servo, perform the following instructions:

1. Set the Goal Position in Dynamixel Wizard to 0, +90 and -90 and observe the position of the arrow on the horn. It should match the picture.

2. Set the Goal Position in Dynamixel Wizard to some position either side of +180 to test if the servo is in joint mode. If it is, it should take the long route around to the second position. If it does not, it is not in joint mode.
   - To convert into joint mode, select "**Joint**" from the dropdown menu for the servo.
     ![Screenshot of Dynamixel Wizard with mode type highlighted using a red box](./images/servo_calibration/to_joint.png)

> If the servo is performing correctly, we can disconnect the servo and use it in a robot.

Reviewer 4: Any error that show up during this process (there should be some sort of error indicator in the status bar of Dynamixel Wizard if I am not mistaken and the LED on the servo will remain lit I think .... its been a while :P) should result in the servo being marked as "bad" and sent off for a more thorough diagnostic (by Peter Turner). Before being flaggeed as "bad" re-flashing the firmware and calibration could be performed a second time just to confirm

Reviewer: I'm gonna make an educated guess on the error indication since I can't test it in person, let me know if it seems close enough lol.

---
section: Guides
chapter: Hardware
title: Servo Setup and Calibration
description: Learn how to test and calibrate servos.
slug: /guides/hardware/servo-calibration
---

Servos are electromechanical components that allow for the precise control and positioning of the various limbs attached to the NUgus robot. Usually, when a servo is found to not be performing correctly (i.e. seizing or a crunching sound can be heard), it should be removed and replaced with a new servo.

> Before a servo is first used, its default parameters need to be updated to ensure correct operation. This is done using [Dynamixel Wizard](https://www.robotis.com/service/download.php?no=1670).

Reviewer 2: Please make it clearer that clicking on this link starts the download and perhaps mention that it's a windows exe

Reviewer: Reworded

## Set Up

To start testing, calibrating or updating the parameters of a servo do the following:

1. Connect the USB2Dynamixel to a USB port on the computer with Dynamixel Wizard.
   ![Photo of the USB2Dynamixel attachment for Dynamixel Wizard connected to a computer](./images/servo_calibration/usb.jpg)

2. Connect the power supply and the servo to the attached component of the USB2Dynamixel.
   ![Photo of the servo cable connected to the power attachment of the USB2Dynamixel](./images/servo_calibration/power_supply_and_comms.jpg)

3. Open Dynamixel Wizard.

## Calibration

To calibrate a servo perform the following instructions:

1. Select "**Tools -> Calibration**". A window will pop up with instructions on how to calibrate the servo.

2. Select the correct servo model. The model will be written on the servo (either MX-64 for head and arm servos or MX-106 for leg servos). Do **NOT** select the 2.0 versions.

> 3. Select the correct communication port from the dropdown menu.

Reviewer 2: Is the correct port obvious?

Reviewer: The correct port will autofill if there's a servo connected

> 4. Disconnect the servo from the USB attachment for a few seconds and then reconnect. If the software does not detect the disconnection, retry. If it fails, redo the calibration from step one. The software will then install the firmware to the servo.

Reviewer 2: We have to install firmware to the servo to calibrate it? Is that correct? That seems strange to me, but I could just be totally out of the loop

Reviewer: The calibration wizard will upload the firmware to the servo during the calibration process, doesn't require user intervention

> 5. Rotate the servo so that the dot is in the 6-o'clock position (as seen in the instructions). Use the tool pictured to move the horn and use the marks on the tool to line up the dot correctly.

Reviewer 2: Where in the instructions? Looking at the preview, this confused me. Could we perhaps have a more specific cross-reference?

Reviewer: There are visuals explaining this section of the calibration within the calibration wizard for users to follow along

3. Update the following parameters (don't forget to press save after each update):

   - ID: Varies depending on where the servo will be placed. See below for guidance.

     | Servo Position       | ID  |
     | -------------------- | --- |
     | Right Shoulder Pitch | 1   |
     | Left Shoulder Pitch  | 2   |
     | Right Shoulder Roll  | 3   |
     | Left Shoulder Roll   | 4   |
     | Right Elbow          | 5   |
     | Left Elbow           | 6   |
     | Right Hip Yaw        | 7   |
     | Left Hip Yaw         | 8   |
     | Right Hip Roll       | 9   |
     | Left Hip Roll        | 10  |
     | Right Hip Pitch      | 11  |
     | Left Hip Pitch       | 12  |
     | Right Knee           | 13  |
     | Left Knee            | 14  |
     | Right Ankle Pitch    | 15  |
     | Left Ankle Pitch     | 16  |
     | Right Ankle Roll     | 17  |
     | Left Ankle Roll      | 18  |
     | Head Yaw             | 19  |
     | Head Pitch           | 20  |

     Note that roll is about the x-axis, pitch is about the y-axis, and yaw is about the z-axis. The below diagram specifies the position of each servo within the robot when it is in its default (zombie) position.

>      <details>

Reviewer 2: Won't this have the same indentation issue?

Reviewer: It seems to work fine for me?

---
section: Guides
chapter: Hardware
title: Servo Setup and Calibration
description: Discover how to calibrate servos.
slug: /guides/hardware/servo-calibration
---

## Servo Calibration

> Servos are electrical components that allow for the precise control and positioning of the various limbs attached to the NUgus robot. Usually, when a servo is found to not be performing correctly (i.e. seizing or a crunching sound can be heard), it is required to be removed and replaced with a new servo. 

Reviewer 3: ```suggestion
Servos are electrical components that allow for the precise control and positioning of the various limbs attached to the NUgus robot. Usually, when a servo is found to not be performing correctly (i.e. seizing or a crunching sound can be heard), it should be removed and replaced with a new servo. 
```

> Servos are electrical components that allow for the precise control and positioning of the various limbs attached to the NUgus robot. Usually, when a servo is found to not be performing correctly (i.e. seizing or a crunching sound can be heard), it is required to be removed and replaced with a new servo. 

Reviewer 4: ```suggestion
Servos are electromechanical components that allow for the precise control and positioning of the various limbs attached to the NUgus robot. Usually, when a servo is found to not be performing correctly (i.e. seizing or a crunching sound can be heard), it should to be removed and replaced with a new servo. 
```

> Before a servo is first used, it\'s default parameters need to be updated to ensure correct operation of the limbs. This is done using [Dynamixel Wizard](https://www.robotis.com/service/download.php?no=1670)

Reviewer 3: ```suggestion
Before a servo is first used, its default parameters need to be updated to ensure correct operation. This is done using [Dynamixel Wizard](https://www.robotis.com/service/download.php?no=1670)
```

To update the servo parameters, perform the following instructions:

> 1. Connect the USB to the computer with the Dynamixel Wizard.

Reviewer 4: ```suggestion
1. Connect the USB2Dynamixel to a USB port on the computer with Dynamixel Wizard.
```

![Photo of the USB attachment for the Dynamixel Wizard connected to a computer](./images/servo_calibration/usb.jpg)

> 2. Connect the power supply and the servo to the attached component of the USB.

Reviewer 4: ```suggestion
2. Connect the power supply and the servo to the attached component of the USB2Dynamixel.
```

![Photo of the servo cable connected to the power attachment of the USB](./images/servo_calibration/power_supply.jpg)

3. Open Dynamixel Wizard

> 4. Select Scan and wait for the attached servo to be discovered
> (OPTIONAL) Brand new servos should be preset to an ID of 1. Lower the ID range in 'Options > Scan' to speed up discovery.

Reviewer 3: ```suggestion
4. Select Scan and wait for the attached servo to be discovered. Optional: brand new servos should be preset to an ID of 1. Lower the ID range in "**Options > Scan**" to speed up discovery.
```

![Screenshot of Dynamixel Wizard performing a scan with no servos discovered](./images/servo_calibration/scan.png)
![Screenshot of Dynamixel Wizard performing a scan with 3 servos discovered](./images/servo_calibration/discover.png)

> 5. If there a multiple servos discovered, the new servo should be the option with ID = 1 and Baud Rate = 57600 bps

Reviewer 3: ```suggestion
5. If there are multiple servos discovered, the new servo should be the option with ID = 1 and Baud Rate = 57600 bps
```

> 6. Update the following parameters (don\'t forget to press save after each update):

Reviewer 3: ```suggestion
6. Update the following parameters (don't forget to press save after each update):
```

  - ID: Varies depending on where the servo will be placed. See below for guidance (TODO)
  - Baud Rate (Bus): 1M bps = 2M/(1 + 1)

  ![Screenshot of Dynamixel Wizard Baud Rate selection with 1M bps selected](./images/servo_calibration/update_parameters.png)

> 7. Test the calibration by setting the Goal Position to 0, +90 and -90 and observing the position of the arrow on the horn.

Reviewer 4: Also ensure that the servo is in joint mode by setting goal position to ~170 degrees and then to ~190 degrees and ensuring that the servo takes the long route during the position change.

Reviewer: I'm assuming that you'd just need to select Joint from the dropdown menu if it's not in Joint mode?

Reviewer 2: @Bidski do you need to manually confirm it as you say, or is the dropdown saying `Joint mode` sufficient?
